src/utilcode/winwrappers.cpp

```diff
@@ -0,0 +1,996 @@
+        _In_ DWORD dwAdditionalFlags
+        )
+{
+    HRESULT hr = S_OK;
+    HANDLE ret = INVALID_HANDLE_VALUE;
+    DWORD lastError;
+    EX_TRY
+    {
+        Win32StackSString path(Win32StackSString::Literal, lpFileName);
+
+        if (SUCCEEDED(LongFile::NormalizePath(path)))
+        {
+            ret = FindFirstFileExW(
+                    path.GetUnicode(),
+                    fInfoLevelId,
+                    lpFindFileData,
+                    fSearchOp,
+                    lpSearchFilter,
+                    dwAdditionalFlags
+                    );
+
+            lastError = GetLastError();
+        }
+    }
+    EX_CATCH_HRESULT(hr);
+
+    if (hr != S_OK )
+    {
+        SetLastError(hr);
+    }
+    else if(ret == INVALID_HANDLE_VALUE)
+    {
+        SetLastError(lastError);
+    }
+
+    return ret;
+}
+#endif //!FEATURE_PAL
+
+#ifndef FEATURE_PAL
+SString LongFile::ExtendedPrefix(SString::Literal, W("\\\\?\\"));
+SString LongFile::DevicePathPrefix(SString::Literal, W("\\\\.\\"));
+SString LongFile::UNCExtendedPathPrefix(SString::Literal, W("\\\\?\\UNC\\"));
+SString LongFile::UNCPathPrefix(SString::Literal, W("\\\\"));
+
+const WCHAR LongFile::VolumeSeparatorChar = W(':');
+const WCHAR LongFile::DirectorySeparatorChar = W('\\');
+const WCHAR LongFile::AltDirectorySeparatorChar = W('/');
+
+BOOL LongFile::IsExtended(SString & path)
+{
+    return path.BeginsWith(ExtendedPrefix);
+}
+
+BOOL LongFile::IsUNCExtended(SString & path)
+{
+
+    return path.BeginsWith(UNCExtendedPathPrefix);
+}
+
+BOOL LongFile::IsDirectorySeparator(WCHAR c)
+{
+    return c == DirectorySeparatorChar || c == AltDirectorySeparatorChar;
+}
+
+BOOL LongFile::IsRelative(SString & path)
+{
+    if (path.GetCount() < 2) 
+    { 
+        return TRUE;  // It isn't fixed, it must be relative.  There is no way to specify a fixed path with one character (or less).
+    }
+
+    if (IsDirectorySeparator(path[0]))
+    {
+        return !IsDirectorySeparator(path[1]); // There is no valid way to specify a relative path with two initial slashes
+    }
+
+    return !((path.GetCount() >= 3)           //The only way to specify a fixed path that doesn't begin with two slashes is the drive, colon, slash format- i.e. C:\
+            && (path[1] == VolumeSeparatorChar)
+            && IsDirectorySeparator(path[2]));
+}
+
+BOOL LongFile::IsDevice(SString & path)
+{
+    return path.BeginsWith(DevicePathPrefix);
+}
+#else
+BOOL LongFile::IsExtended(SString & path)
+{
+    return FALSE;
+}
+
+BOOL LongFile::IsUNCExtended(SString & path)
+{
+    return FALSE;
+}
+
+BOOL LongFile::IsDirectorySeparator(WCHAR c)
+{
+    return FALSE;
+}
+
+BOOL LongFile::IsRelative(SString & path)
+{
+    return FALSE;
+}
+
+BOOL LongFile::IsDevice(SString & path)
+{
+    return FALSE;
+}
+#endif //FEATURE_PAL
+
+HRESULT LongFile::NormalizePath(SString & path)
+{
+    HRESULT hr = S_OK;
+#ifndef FEATURE_PAL
+
+    if (path.IsEmpty() || IsExtended(path) || IsUNCExtended(path))
```

Comment:
I dont believe this method should be invoked when path is empty. Is there a scenario for it?